nistats/second_level_model.py

```diff
@@ -0,0 +1,451 @@
+                raise ValueError('first_level_conditions_name must be a '
+                                 'list of strings')
+            if len(first_level_conditions_name) != len(first_level_conditions):
+                raise ValueError('length of first_level_conditions_name and'
+                                 ' first_level_conditions must match')
+            for name in first_level_conditions_name:
+                if not isinstance(name, str):
+                    raise ValueError('first_level_conditions_name must be a '
+                                     'list of strings')
+
+        # Build the design matrix X and list of imgs Y for GLM fit
+        if isinstance(first_level_input, pd.DataFrame):
+            # Get only first level conditions if provided
+            if first_level_conditions is not None:
+                for condition in first_level_conditions:
+                    if condition not in first_level_input['map_name'].tolist():
+                        raise ValueError('condition %s not present in'
+                                         ' first_level_input' % condition)
+                in_cond = first_level_input.apply(lambda x: x['map_name'] in
+                                                  first_level_conditions,
+                                                  axis=1)
+                first_level_input = first_level_input[in_cond]
+            # Create design matrix if necessary
+            if design_matrix is None:
+                design = _create_second_level_design(first_level_input,
+                                                     confounds)
+            # get effect maps
+            effects_maps = first_level_input['effects_map_path'].tolist()
+
+        elif isinstance(first_level_input[0], FirstLevelModel):
+            # Check models were fit
+            for midx, model in enumerate(first_level_input):
+                if model.labels_ is None:
+                    raise ValueError('Model at idx %d has not been fit' % midx)
+            # Get the first level model maps
+            design_df = pd.DataFrame(columns=['map_name', 'subject_id'])
+            effects_maps = []
+            for model in first_level_input:
+                for con_idx, con_def in enumerate(first_level_conditions):
+                    con_name = 'con_%02d' % con_idx
+                    if first_level_conditions_name:
+                        con_name = first_level_conditions_name[con_idx]
+                    design_df.loc[len(design_df)] = [con_name,
+                                                     model.subject_id]
+                    eff_map = model.compute_contrast(con_def,
+                                                     output_type='effect_size')
+                    effects_maps.append(eff_map)
+            # Get the design matrix
+            if design_matrix is None:
+                design = _create_second_level_design(design_df, confounds)
```

Comment:
I find it hard to undertsand the distinction between design_df and design_matrix.
* I think that the design of _create_second_levle _design is a key issue. This function should be part of the public API, and there has to be an example dedicated only to presenting what it does.
* Renaming will be important. `design` alone is not clear enough; `design_matrix` is certainly better. `design_df` is not great; may be `design_input`